src/main/java/org/commcare/formplayer/installers/FormplayerLocaleInstaller.java

```diff
@@ -0,0 +1,102 @@
+import org.javarosa.core.services.locale.TableLocaleSource;
+import org.javarosa.core.util.externalizable.DeserializationException;
+import org.javarosa.core.util.externalizable.ExtUtil;
+import org.javarosa.core.util.externalizable.ExtWrapMap;
+import org.javarosa.core.util.externalizable.PrototypeFactory;
+import org.javarosa.xml.util.InvalidStructureException;
+import org.javarosa.xml.util.UnfullfilledRequirementsException;
+import org.xmlpull.v1.XmlPullParserException;
+
+import java.io.DataInputStream;
+import java.io.DataOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Hashtable;
+import java.util.Vector;
+
+/**
+ * ResourceInstaller for locale files that always persists the locale data in
+ * the SQLite DB.
+ */
+public class FormplayerLocaleInstaller extends SimpleInstaller {
+
+    private String locale;
+
+    private Hashtable<String, String> localizedValues;
+
+    /**
+     * Serialization only!
+     */
+    @SuppressWarnings("unused")
+    public FormplayerLocaleInstaller() {
+    }
+
+    public FormplayerLocaleInstaller(String locale) {
+        this.locale = locale;
+    }
+
+    @Override
+    public boolean initialize(CommCarePlatform platform, boolean isUpgrade) throws IOException, InvalidReferenceException, InvalidStructureException, XmlPullParserException, UnfullfilledRequirementsException {
+        Localization.getGlobalLocalizerAdvanced().addAvailableLocale(locale);
+        Localization.getGlobalLocalizerAdvanced().registerLocaleResource(locale, new TableLocaleSource(localizedValues));
+        return true;
+    }
+
+    @Override
+    public boolean install(Resource r, ResourceLocation location, Reference ref, ResourceTable table, CommCarePlatform platform, boolean upgrade, ResourceInstallContext resourceInstallContext) throws UnresolvedResourceException {
+        try {
+            try (InputStream incoming = ref.getStream()) {
+                localizedValues = LocalizationUtils.parseLocaleInput(incoming);
+                table.commit(r, upgrade ? Resource.RESOURCE_STATUS_UPGRADE : Resource.RESOURCE_STATUS_INSTALLED);
+            }
+            return true;
+        } catch (IOException e) {
+            throw new UnreliableSourceException(r, e.getMessage());
+        }
+    }
+
+    @Override
+    public boolean uninstall(Resource r, CommCarePlatform platform) throws UnresolvedResourceException {
+        if (localizedValues != null) {
+            localizedValues.clear();
+            localizedValues = null;
+        }
+        return true;
+    }
+
+    @Override
+    public boolean verifyInstallation(Resource r, Vector<MissingMediaException> problemList, CommCarePlatform platform) {
+        if (locale == null) {
+            problemList.addElement(new MissingMediaException(r, "Bad metadata, no locale",
+                    MissingMediaException.MissingMediaExceptionType.NONE));
+        }
+        if (localizedValues == null) {
```

Comment:
not actually sure if this should be checked since and empty hashmap would end up being null

Reply:
Agreed. `verifyInstallation` can presume that the record is fully installed, and an empty/null table or null locale would only be possible if the record was malformed, so it would be better for those to fail to install, rather than error during the check.